lib/ansible/galaxy/dependency_resolution/reporters.py

```diff
@@ -5,16 +5,103 @@
 
 from __future__ import annotations
 
+import typing as t
+from collections import defaultdict
+from contextlib import suppress as suppress_exceptions
+
 try:
     from resolvelib import BaseReporter
 except ImportError:
     class BaseReporter:  # type: ignore[no-redef]
         pass
 
+try:
+    from resolvelib.resolvers import Criterion
+except ImportError:
+    try:
+        from resolvelib.resolvers.criterion import Criterion
+    except ImportError:
+        class Criterion:  # type: ignore[no-redef]
+            pass
+
+from ansible.utils.display import Display
+from .dataclasses import Candidate, Requirement
+
+
+display = Display()
+
+
+_CLI_APP_NAME = 'ansible-galaxy'
+_MESSAGES_AT_REJECT_COUNT = {
+    1: (
+        f'{_CLI_APP_NAME} is looking at multiple versions of {{fqcn}} to '
+        'determine which version is compatible with other '
+        'requirements. This could take a while.'
+    ),
+    8: (
+        f'{_CLI_APP_NAME} is looking at multiple versions of {{fqcn}} to '
+        'determine which version is compatible with other '
+        'requirements. This could take a while.'
+    ),
+    13: (
+        'This is taking longer than usual. You might need to provide '
+        'the dependency resolver with stricter constraints to reduce '
+        'runtime. If you want to abort this run, press Ctrl + C.'
+    ),
+}
+
 
 class CollectionDependencyReporter(BaseReporter):
     """A dependency reporter for Ansible Collections.
 
     This is a proxy class allowing us to abstract away importing resolvelib
     outside of the `ansible.galaxy.dependency_resolution` Python package.
     """
+
+    def __init__(self) -> None:
+        """Initialize the collection rejection counter."""
+        super().__init__()
+
+        self.reject_count_by_fqcn: t.DefaultDict[str, int] = defaultdict(int)
```

Comment:
Can we use the `defaultdict[str, int]`  as the type annotation and not the deprecated typing one.

Reply:
I didn't realize it's deprecated. Also, I started this in 2023, and it might not have been possible back then.

I'll give it a try, thanks!

Reply:
You're right. I'll add the suggested annotation. Marking this thread as resolved.